src/python/pants/binaries/binary_tool.py

```diff
@@ -230,3 +297,113 @@ def tar_xz_extractor(self):
 
   def _executable_location(self):
     return os.path.join(self.select(), 'bin', 'xz')
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class VersionDigestMapping:
+  """Parse the --version-digest-mapping option back into a dictionary."""
+  version_digest_mapping: Tuple[Tuple[str, Tuple[str, str, int]], ...]
+
+  def __init__(self, version_digest_mapping: Dict[str, List[Union[str, int]]]) -> None:
+    self.version_digest_mapping = tuple(
+      (platform_constraint, tuple(data)) # type: ignore[misc]
+      for platform_constraint, data in version_digest_mapping.items()
+    )
+
+  @memoized_property
+  def _deserialized_mapping(
+      self,
+  ) -> Dict[PlatformConstraint, ToolForPlatform]:
+    deserialized: Dict[PlatformConstraint, ToolForPlatform] = {}
+    for platform_constraint, (version, fingerprint, size_bytes) in self.version_digest_mapping:
+      deserialized[PlatformConstraint(platform_constraint)] = ToolForPlatform(
+        version=ToolVersion(version),
+        digest=Digest(fingerprint, size_bytes),
+      )
+    return deserialized
+
+  def get(self, platform_constraint: PlatformConstraint) -> ToolForPlatform:
+    return self._deserialized_mapping[platform_constraint]
+
+
+@dataclass(frozen=True)
+class BinaryToolUrlSet:
+  tool_for_platform: ToolForPlatform
+  host_platform: HostPlatform
+  url_generator: BinaryToolUrlGenerator
+
+  def get_urls(self) -> List[str]:
+    return self.url_generator.generate_urls(
+      version=self.tool_for_platform.version.version,
+      host_platform=self.host_platform if self.host_platform != HostPlatform.empty else None)
+
+
+@frozen_after_init
```

Comment:
Is there a reference somewhere explaining why we use the `frozen_after_init`/`unsafe_hash` combo rather than overriding `__new__`, or using a factory function?

Reply:
No -- and I like overriding `__new__`! It feels nice to return a value. @Eric-Arellano do you have any further context on this?

Reply:
> rather than overriding __new__

I think `__new__` will break dataclasses. When we came up with `@frozen_after_init`, we wanted some way to have custom `__init__()`s but keep the immutability of `frozen=True` and the hashing of dataclasses. The decorator allows us to do all three things.